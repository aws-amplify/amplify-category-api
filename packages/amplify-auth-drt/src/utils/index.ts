export * from './internal-representation';
export * from './test-utils';
