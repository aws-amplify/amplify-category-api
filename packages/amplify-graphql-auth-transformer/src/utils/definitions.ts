import { GetArgumentsOptions } from '@aws-amplify/graphql-transformer-core';
import { AuthDirective } from '@aws-amplify/graphql-directives';

/**
 * AuthStrategy
 */
export type AuthStrategy = 'owner' | 'groups' | 'public' | 'private' | 'custom';
/**
 * AuthProvider
 */
export type AuthProvider = 'apiKey' | 'iam' | 'identityPool' | 'oidc' | 'userPools' | 'function';
/**
 * ModelMutation
 */
export type ModelMutation = 'create' | 'update' | 'delete';
/**
 * ModelOperation
 */
export type ModelOperation = 'create' | 'update' | 'delete' | 'get' | 'list' | 'sync' | 'search' | 'listen';

/**
 * RelationalPrimaryMapConfig
 */
export type RelationalPrimaryMapConfig = Map<string, { claim: string; field: string }>;
/**
 * SearchableConfig
 */
export interface SearchableConfig {
  queries: {
    search: string;
  };
}

export type GetAuthRulesOptions = GetArgumentsOptions & {
  isField?: boolean;
  isSqlDataSource?: boolean;
};

/**
 * AuthRule
 */
export interface AuthRule {
  allow: AuthStrategy;
  provider?: AuthProvider;
  ownerField?: string;
  identityClaim?: string;
  groupsField?: string;
  groupClaim?: string;
  groups?: string[];
  operations?: ModelOperation[];
  // Used only for IAM provider to decide if an IAM policy needs to be generated. IAM auth with AdminUI does not need IAM policies
  generateIAMPolicy?: boolean;
}

/**
 * AuthDirective
 */
export interface AuthDirective {
  rules: AuthRule[];
}

/**
 * ConfiguredAuthProviders
 */
export interface ConfiguredAuthProviders {
  default: AuthProvider;
  onlyDefaultAuthProviderConfigured: boolean;
  hasApiKey: boolean;
  hasUserPools: boolean;
  hasOIDC: boolean;
  hasIAM: boolean;
  hasLambda: boolean;
  hasAdminRolesEnabled: boolean;
  hasIdentityPoolId: boolean;
  shouldAddDefaultServiceDirective: boolean;
  genericIamAccessEnabled: boolean;
}

export const authDirectiveDefinition = AuthDirective.definition;
