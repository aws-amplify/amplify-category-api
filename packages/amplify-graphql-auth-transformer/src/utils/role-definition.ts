import { AuthProvider, AuthStrategy } from './definitions';

/**
 * RoleDefinition
 */
export interface RoleDefinition {
  provider: AuthProvider;
  strategy: AuthStrategy;
  static: boolean;
  claim?: string;
  entity?: string;
  // specific to mutations
  allowedFields?: Array<string>;
  nullAllowedFields?: Array<string>;
  areAllFieldsAllowed?: boolean;
  areAllFieldsNullAllowed?: boolean;
  isEntityList?: boolean;
}

/**
 * RolesByProvider
 */
export interface RolesByProvider {
  cognitoStaticRoles: Array<RoleDefinition>;
  cognitoDynamicRoles: Array<RoleDefinition>;
  oidcStaticRoles: Array<RoleDefinition>;
  oidcDynamicRoles: Array<RoleDefinition>;
  iamRoles: Array<RoleDefinition>;
  apiKeyRoles: Array<RoleDefinition>;
  lambdaRoles: Array<RoleDefinition>;
}

/**
 * Compare if two auth roles are identical
 * @param role1 auth role 1
 * @param role2 auth role 2
 * @returns boolean for two roles are identical
 */
export const isIdenticalAuthRole = (role1: RoleDefinition, role2: RoleDefinition): boolean =>
  role1.provider === role2.provider &&
  role1.strategy === role2.strategy &&
  role1.static === role2.static &&
  role1.claim === role2.claim &&
  role1.entity === role2.entity;

/**
 * Determine if the given relational field role have the same access as one of the related model auth roles
 * Usually this means an auth role have the exact same provider, strategy, claim and entity
 * One special case is when the provider 'userPools', 'oidc', or 'identityPool' exists on both sides and at least one side have 'private' role
 * @param fieldRole relational field auth role
 * @param relatedModelRoles related model auth roles
 * @returns boolean for the field can access both sides or not
 */
export const isFieldRoleHavingAccessToBothSide = (fieldRole: RoleDefinition, relatedModelRoles: RoleDefinition[]): boolean =>
  relatedModelRoles.some(
    (relatedRole) =>
      isIdenticalAuthRole(relatedRole, fieldRole) ||
      (relatedRole.provider === fieldRole.provider &&
        (relatedRole.provider === 'userPools' || relatedRole.provider === 'oidc' || relatedRole.provider === 'identityPool') &&
        (fieldRole.strategy === 'private' || relatedRole.strategy === 'private')),
  );

/**
 * Determine the given auth role is either dynamic auth role or custom auth role
 * These auth client authorization depends on the dynamic field or custom auth logic apart from the auth type
 * Which authenticates differently based on the given dynamic field value
 * @param role auth role definition
 * @returns boolean for the auth role is a dynamic or custom auth role
 */
export const isDynamicAuthOrCustomAuth = (role: RoleDefinition): boolean =>
  (role.static === false && (role.strategy === 'owner' || role.strategy === 'groups')) || role.strategy === 'custom';
