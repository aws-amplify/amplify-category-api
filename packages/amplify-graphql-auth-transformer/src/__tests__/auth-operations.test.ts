import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { ConflictHandlerType, GraphQLTransform, SyncConfig } from '@aws-amplify/graphql-transformer-core';
import { AppSyncAuthConfiguration } from '@aws-amplify/graphql-transformer-interfaces';
import { SearchableModelTransformer } from '@aws-amplify/graphql-searchable-transformer';
import { AccessControlMatrix } from '../accesscontrol';
import { AuthTransformer } from '../graphql-auth-transformer';
import { MODEL_OPERATIONS } from '../utils';

test('invalid granular read operation at the field level', () => {
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [{
      authenticationType: 'API_KEY',
    }],
  };
  const invalidSchema = `
    type Test @model @auth(rules: [{ allow: public, operations: [ list, create ]}]) {
      id: ID!
      name: String @auth(rules: [{ allow: public, operations: [ get, create ]}])
    }`;
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  expect(() => transformer.transform(invalidSchema)).toThrowError(
    '\'get\' operation is not allowed at the field level.',
  );
});

test('renamed subscriptions should generate auth resolver', () => {
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [{
      authenticationType: 'API_KEY',
    }],
  };
  const validSchema = `
    type Note
      @model(subscriptions: { onUpdate: ["onNoteUpdate"] })
      @auth(rules: [{ allow: private, provider: userPools }])
    {
      noteId: String!
      userId: String!
      assignedTo: String!
      comments: String
    }`;
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();

  // should generate auth resolver for renamed subscription
  expect(out.resolvers['Subscription.onNoteUpdate.auth.1.req.vtl']).toBeDefined();
  expect(out.resolvers['Subscription.onNoteUpdate.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onNoteUpdate.postAuth.1.req.vtl']).toBeDefined();
  expect(out.resolvers['Subscription.onNoteUpdate.postAuth.1.req.vtl']).toMatchSnapshot();
});

test('invalid read list operation combination', () => {
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [],
  };
  const invalidSchema = `
    type Test @model @auth(rules: [{ allow: public, operations: [ read, list, create ]}]) {
      id: ID!
      name: String
    }`;
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  expect(() => transformer.transform(invalidSchema)).toThrowError(
    '\'list\' operations are specified in addition to \'read\'. Either remove \'read\' to limit access only to \'list\' or only keep \'read\' to grant all get,list,search,listen,sync access.',
  );
});

test('invalid read get operation combination', () => {
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [],
  };
  const invalidSchema = `
    type Test @model @auth(rules: [{ allow: public, operations: [ read, get, create ]}]) {
      id: ID!
      name: String
    }`;
  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new AuthTransformer()],
  });
  expect(() => transformer.transform(invalidSchema)).toThrowError(
    '\'get\' operations are specified in addition to \'read\'. Either remove \'read\' to limit access only to \'get\' or only keep \'read\' to grant all get,list,search,listen,sync access.',
  );
});

test('read access control', () => {
  /*
  given the following schema
  type TestList @model @auth(rules: [{ allow: public, operations: [ get, list, create ]}]) {
    id: ID!
    name: String
  }
  */

  const ownerRole = 'userPools:owner:id';
  const typeFields = ['id', 'name'];

  const acm = new AccessControlMatrix({
    name: 'TestList',
    resources: typeFields,
    operations: MODEL_OPERATIONS,
  });

  acm.setRole({
    role: ownerRole,
    operations: ['get', 'list'],
  });

  typeFields.forEach((field) => {
    expect(acm.isAllowed(ownerRole, field, 'list')).toBe(true);
    expect(acm.isAllowed(ownerRole, field, 'get')).toBe(true);
  });
});

test('list access control', () => {
  /*
  given the following schema
  type TestList @model @auth(rules: [{ allow: public, operations: [ list, create ]}]) {
    id: ID!
    name: String
  }
  */

  const ownerRole = 'userPools:owner:id';
  const typeFields = ['id', 'name'];

  const acm = new AccessControlMatrix({
    name: 'TestList',
    resources: typeFields,
    operations: MODEL_OPERATIONS,
  });

  acm.setRole({
    role: ownerRole,
    operations: ['list'],
  });

  typeFields.forEach((field) => {
    expect(acm.isAllowed(ownerRole, field, 'list')).toBe(true);
    expect(acm.isAllowed(ownerRole, field, 'get')).toBe(false);
  });
});

test('get access control', () => {
  /*
  given the following schema
  type TestList @model @auth(rules: [{ allow: public, operations: [ get, create ]}]) {
    id: ID!
    name: String
  }
  */

  const ownerRole = 'userPools:owner:id';
  const typeFields = ['id', 'name'];

  const acm = new AccessControlMatrix({
    name: 'TestList',
    resources: typeFields,
    operations: MODEL_OPERATIONS,
  });

  acm.setRole({
    role: ownerRole,
    operations: ['get'],
  });

  typeFields.forEach((field) => {
    expect(acm.isAllowed(ownerRole, field, 'list')).toBe(false);
    expect(acm.isAllowed(ownerRole, field, 'get')).toBe(true);
  });
});

test('read get list auth operations', () => {
  const config: SyncConfig = {
    ConflictDetection: 'VERSION',
    ConflictHandler: ConflictHandlerType.AUTOMERGE,
  };

  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'API_KEY',
    },
    additionalAuthenticationProviders: [],
  };
  const validSchema = `
    type TestSubscription @model @searchable @auth(rules: [{ allow: public, operations: [ listen ]}]) {
      id: ID!
      name: String
    }
    type TestSearch @model @searchable @auth(rules: [{ allow: public, operations: [ search ]}]) {
      id: ID!
      name: String
    }
    type TestSync @model @searchable @auth(rules: [{ allow: public, operations: [ sync ]}]) {
      id: ID!
      name: String
    }
    type TestList @model @searchable @auth(rules: [{ allow: public, operations: [ list ]}]) {
      id: ID!
      name: String
    }
    type TestGet @model @searchable @auth(rules: [{ allow: public, operations: [ get ]}]) {
      id: ID!
      name: String
    }
    type TestRead @model @searchable @auth(rules: [{ allow: public, operations: [ read ]}]) {
      id: ID!
      name: String
    }
  `;

  const transformer = new GraphQLTransform({
    authConfig,
    transformers: [new ModelTransformer(), new SearchableModelTransformer(), new AuthTransformer()],
    resolverConfig: {
      project: config,
    },
  });

  const out = transformer.transform(validSchema);
  expect(out).toBeDefined();

  // listen
  expect(out.resolvers['Query.getTestSubscription.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.listTestSubscriptions.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onCreateTestSubscription.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onDeleteTestSubscription.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onUpdateTestSubscription.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.searchTestSubscriptions.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.syncTestSubscriptions.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');

  // search
  expect(out.resolvers['Query.getTestSearch.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.listTestSearches.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onCreateTestSearch.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onDeleteTestSearch.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onUpdateTestSearch.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.searchTestSearches.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.syncTestSearches.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');

  // sync
  expect(out.resolvers['Query.getTestSync.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.listTestSyncs.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onCreateTestSync.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onDeleteTestSync.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onUpdateTestSync.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.searchTestSyncs.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.syncTestSyncs.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');

  // get
  expect(out.resolvers['Query.getTestList.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.listTestLists.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onCreateTestList.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onDeleteTestList.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onUpdateTestList.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.getTestList.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.listTestLists.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onCreateTestList.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onDeleteTestList.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onUpdateTestList.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');

  // list
  expect(out.resolvers['Query.getTestGet.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.listTestGets.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onCreateTestGet.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onDeleteTestGet.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onUpdateTestGet.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.getTestGet.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.listTestGets.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onCreateTestGet.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onDeleteTestGet.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onUpdateTestGet.auth.1.req.vtl']).not.toContain('#set( $isAuthorized = true )');

  // read
  expect(out.resolvers['Query.getTestRead.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.listTestReads.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onCreateTestRead.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onDeleteTestRead.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Subscription.onUpdateTestRead.auth.1.req.vtl']).toMatchSnapshot();
  expect(out.resolvers['Query.getTestRead.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Query.listTestReads.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onCreateTestRead.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onDeleteTestRead.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
  expect(out.resolvers['Subscription.onUpdateTestRead.auth.1.req.vtl']).toContain('#set( $isAuthorized = true )');
});
