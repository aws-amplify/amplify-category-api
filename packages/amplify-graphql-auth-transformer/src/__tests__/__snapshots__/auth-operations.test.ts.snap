// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can update on model without delete 1`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#set( $nullAllowedFields = [] )
#set( $deniedFields = {} )
#if( $util.authType() == \\"API Key Authorization\\" )
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #if( !$isAuthorized )
    #set( $ownerEntity0 = $util.defaultIfNull($ctx.result.owner, null) )
    #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get(\\"sub\\"), null) )
    #set( $currentClaim1 = $util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), null)) )
    #if( !$util.isNull($ownerClaim0) && !$util.isNull($currentClaim1) )
      #set( $ownerClaim0 = \\"$ownerClaim0::$currentClaim1\\" )
      #set( $ownerClaimsList0 = [] )
      $util.qr($ownerClaimsList0.add($util.defaultIfNull($ctx.identity.claims.get(\\"sub\\"), null)))
      $util.qr($ownerClaimsList0.add($util.defaultIfNull($ctx.identity.claims.get(\\"username\\"), $util.defaultIfNull($ctx.identity.claims.get(\\"cognito:username\\"), null))))
      #set( $ownerAllowedFields0 = [\\"email\\",\\"nick\\",\\"id\\",\\"_version\\",\\"_deleted\\",\\"_lastChangedAt\\"] )
      #set( $ownerNullAllowedFields0 = [] )
      #set( $isAuthorizedOnAllFields0 = false )
      #if( $ownerEntity0 == $ownerClaim0 || $ownerClaimsList0.contains($ownerEntity0) )
        #if( $isAuthorizedOnAllFields0 )
          #set( $isAuthorized = true )
        #else
          $util.qr($allowedFields.addAll($ownerAllowedFields0))
          $util.qr($nullAllowedFields.addAll($ownerNullAllowedFields0))
        #end
      #end
    #end
  #end
#end
#if( !$isAuthorized && $allowedFields.isEmpty() && $nullAllowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #foreach( $entry in $util.map.copyAndRetainAllKeys($ctx.args.input, $inputFields).entrySet() )
    #if( $util.isNull($entry.value) && !$nullAllowedFields.contains($entry.key) )
      $util.qr($deniedFields.put($entry.key, \\"\\"))
    #end
  #end
  #foreach( $deniedField in $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
    $util.qr($deniedFields.put($deniedField, \\"\\"))
  #end
#end
#if( $deniedFields.keySet().size() > 0 )
  $util.error(\\"Unauthorized on \${deniedFields.keySet()}\\", \\"Unauthorized\\")
#end
$util.toJson({})
## [End] Authorization Steps. **"
`;

exports[`can update on model without delete 2`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )
$util.unauthorized()
#end
#if( $util.authType() == \\"User Pool Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 1`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 2`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == \\"API Key Authorization\\" )
  #set( $isAuthorized = true )
#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 3`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 4`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 5`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 6`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == \\"API Key Authorization\\" )
  #set( $isAuthorized = true )
#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 7`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 8`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 9`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 10`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 11`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == \\"API Key Authorization\\" )
  #set( $isAuthorized = true )
#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 12`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#set( $primaryFieldMap = {} )
#if( $util.authType() == \\"API Key Authorization\\" )
  #set( $isAuthorized = true )
#end
#if( !$isAuthorized && $util.isNull($ctx.stash.authFilter) )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 13`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )
  #set( $isAuthorized = true )
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 14`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )
  #set( $isAuthorized = true )
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`read get list auth operations 15`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )
  #set( $isAuthorized = true )
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`renamed subscriptions should generate auth resolver 1`] = `
"## [Start] Authorization Steps. **
$util.qr($ctx.stash.put(\\"hasAuth\\", true))
#set( $isAuthorized = false )
#if( $util.authType() == \\"API Key Authorization\\" )

#end
#if( $util.authType() == \\"User Pool Authorization\\" )
  #set( $isAuthorized = true )
  #set( $hasValidOwnerArgument = false )
  #set( $authRuntimeFilter = [] )
  #set( $authOwnerRuntimeFilter = [] )
  #set( $authGroupRuntimeFilter = [] )
  ## Apply dynamic roles auth if not previously authorized by static groups and owner argument **
  #if( $authOwnerRuntimeFilter.size() > 0 )
    $util.qr($authRuntimeFilter.addAll($authOwnerRuntimeFilter))
  #end
  #if( $authGroupRuntimeFilter.size() > 0 )
    $util.qr($authRuntimeFilter.addAll($authGroupRuntimeFilter))
  #end
  #set( $filterArgsSize = 0 )
  #if( !$util.isNullOrEmpty($ctx.args.filter) )
    #set( $filterArgsSize = $ctx.args.filter.size() )
  #end
  #set( $isOwnerAuthAuthorizedAndNoOtherFilters = $hasValidOwnerArgument && $authRuntimeFilter.size() == 1 && $filterArgsSize == 0 )
  #set( $isOwnerOrDynamicAuthAuthorizedWithFilters = (!$isAuthorized || $hasValidOwnerArgument) && $authRuntimeFilter.size() > 0 )
  #if( !$isOwnerAuthAuthorizedAndNoOtherFilters && $isOwnerOrDynamicAuthAuthorizedWithFilters )
    #if( $util.isNullOrEmpty($ctx.args.filter) )
      #set( $ctx.args.filter = { \\"or\\": $authRuntimeFilter } )
    #else
      #set( $ctx.args.filter = { \\"and\\": [ { \\"or\\": $authRuntimeFilter }, $ctx.args.filter ]} )
    #end
    #set( $isAuthorized = true )
  #end
#end
#if( !$isAuthorized )
$util.unauthorized()
#end
$util.toJson({\\"version\\":\\"2018-05-29\\",\\"payload\\":{}})
## [End] Authorization Steps. **"
`;

exports[`renamed subscriptions should generate auth resolver 2`] = `
"## [Start] Sandbox Mode Disabled, IAM Access Disabled. **
#if( !$ctx.stash.get(\\"hasAuth\\") )
  $util.unauthorized()
#end
$util.toJson({})
## [End] Sandbox Mode Disabled, IAM Access Disabled. **"
`;
